File: app/src/main/java/com/albineli/udacity/popularmovies/repository/RepositoryBase.java
package com.albineli.udacity.popularmovies.repository;


import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;

public abstract class RepositoryBase {
    protected <T> Observable<T> observeOnMainThread(Observable<T> observable) {
        return observable.observeOn(AndroidSchedulers.mainThread());
    }

    protected  Completable observeOnMainThread(Completable completable) {
        return completable.observeOn(AndroidSchedulers.mainThread());
    }
}

File: app/src/main/java/com/themovielist/repository/RepositoryBase.kt
package com.themovielist.repository


import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import retrofit2.Retrofit

abstract class RepositoryBase<T> constructor(private val retrofit: Retrofit) {
    abstract val getApiInstanceType: Class<T>

    protected val mApiInstance: T by lazy { retrofit.create(getApiInstanceType) }

    protected fun <T> observeOnMainThread(observable: Observable<T>): Observable<T> {
        return observable.observeOn(AndroidSchedulers.mainThread())
    }

    protected fun observeOnMainThread(completable: Completable): Completable {
        return completable.observeOn(AndroidSchedulers.mainThread())
    }
}
